Add shuffleCows options for randomizer

garrettjoecox opened the pull request:
## Feedback wanted!

This is an initial attempt at getting the shuffleCows option working in randomizer. I did this without a ton of context so please let me know if this is in any way not consistent with the other options/rest of the codebase.

Summary of changes:
- Add cow checks in `randomizer.cpp`
- Add Cow Shuffle option on GUI
- Move two cows, one in the stable (it is too close to the other cow) and the other in the lon lon tower (was blocking the other cow)
- Using `cowsMilked` global array to store check state
- If the player has already received the item it will act like vanilla (milk if empty bottle, sorry message if no empty bottle)

Could be better/TODO:
- I don't love using the X + Z coordinate combo to identify the cows, but with 2 cows in a few areas it's the best I could come up with.
- [x] ~It isn't clear to me at all why there are two funcs (func_809DF778,func_809DF7D8)  that originally give the milk item, so for now I altered both of them. Probably need to investigate this before merging~ This seems to be some sort of common failsafe for some item types, which it actually helps me out here for ice traps.
- [x] Determine if I need a player flag to protect from blocking check (related PR https://github.com/HarbourMasters/Shipwright/pull/961)
- [x] Ensure ice traps work
- [x] cowMilked status not persisting in save
- [x] Fix bug with moved cow's tail being offset
- [ ] Move functions out of cow actor file?

New cow location in lon lon tower
<img width="1066" alt="Screen Shot 2022-08-03 at 12 13 36 AM" src="https://user-images.githubusercontent.com/7316699/182530486-8538efec-dc1d-4236-a9b2-b591f19292d9.png">

New cow location in stable
<img width="1066" alt="Screen Shot 2022-08-03 at 12 19 07 AM" src="https://user-images.githubusercontent.com/7316699/182530528-761cbfe9-382e-4e80-b798-66d43af6da9f.png">

Attempting to obtain an item again + without an empty bottle
<img width="1066" alt="Screen Shot 2022-08-03 at 12 14 45 AM" src="https://user-images.githubusercontent.com/7316699/182530674-ff3565f7-18a2-498c-b91c-5edff47ed379.png">

garrettjoecox committed:
Add shuffleCows options for randomizer

garrettjoecox commented:
Closes #730

leggettc18 commented:
Haven't looked too much at the code yet, but I like what I'm seeing in these screenshots. A few things to test. Hold R with a shield equipped while playing Epona's song, and if the check gets skipped, you'll need to implement the pending flags stuff from #961. Make sure Ice Traps work normally, I think they should be fine but if any item is going to cause issues it's them.

garrettjoecox commented:
> Hold R with a shield equipped while playing Epona's song, and if the check gets skipped, you'll need to implement the pending flags stuff

Not sure I fully understand this, might need to see a demonstration but from what I can tell it's working

> Make sure Ice Traps work normally

Ice traps are broken I think, just a sound and damage animation but no ice block, no damage taken and no message box

garrettjoecox commented:
> Ice traps are broken I think, just a sound and damage animation but no ice block, no damage taken and no message box

~It would appear this has something to do with the two funcs giving milk in vanilla, removing one of them fixes ice traps 🤔 , will investigate more later~

Not as simple as removing one of them it has unwanted side-effects

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -106,6 +111,18 @@ void EnCow_Init(Actor* thisx, GlobalContext* globalCtx) {
     s32 pad;
 
+    // Move left cow in lon lon tower
+    if (EnCow_IsSecondCowInTower(thisx, globalCtx)) {
+        this->actor.world.pos.x = -229.0f;
+        this->actor.world.pos.z = 157.0f;
+        this->actor.shape.rot.y = 15783.0f;
+    }
```
are these cows moved in 3drando and/or 64rando?

garrettjoecox replied:
Not sure about 3drando but yes for 64rando

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +226,57 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
 
+bool EnCow_IsSecondCowInTower(EnCow* this, GlobalContext* globalCtx) {
+    return 
+        gSaveContext.n64ddFlag && 
```
does it make sense to have the rando check in here? the cow is still the second cow in the tower even if we're not randomized, so it seems to me it would make more sense to keep those checks separate

so instead of having
```c
if (EnCow_IsSecondCowInTower(thisx, globalCtx)) {
```
we'd have
```c
if (gSaveContext.n64ddFlag && EnCow_IsSecondCowInTower(thisx, globalCtx)) {
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +226,57 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+    return 
+        gSaveContext.n64ddFlag && 
+        globalCtx->sceneNum == SCENE_SOUKO && 
+        (this->actor.world.pos.x + this->actor.world.pos.z == -173 || this->actor.world.pos.x + this->actor.world.pos.z == -72);
```
using x+z seems really odd here, is there a reason to do that instead of just checking against known x,y,z coords?

aMannus replied:
As said a little bit below here, this should also check for the actor params instead of position values.

leggettc18 replied:
I think he confirmed both cows have the same actor params. In vanilla besides their positions these cows are essentially identical.

garrettjoecox replied:
Indeed, actorParams for all cows is the same, from what I can tell each cow has two actors, one is the visual cow with actorParam 0 and the second is the collision box with actorParam 1.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +226,57 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+GetItemID EnCow_GetItemFromCow(EnCow* this, GlobalContext* globalCtx, bool setFlag) {
+    GetItemID itemId = ITEM_NONE;
+
+    if (Inventory_HasEmptyBottle()) {
+        return GI_MILK;
+    }
```
it seems like having this here will make it so milk overrides the rando check if the player has an empty bottle

garrettjoecox replied:
This runs first, then the rando check after so the rando would take priority, but I’ll move this to an else to make it more readable

garrettjoecox replied:
Oops actually you're right, this was meant to be `itemId = GI_MILK;` not return, regardless It's moving 👍

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +226,57 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+        return GI_MILK;
+    }
+
+    s32 collectableFlag = 0x1A;
+    if (EnCow_IsSecondCow(this, globalCtx)) {
+        collectableFlag = 0x19;
```
just want to verify we're not using these flags for anything else in the scenes with cows

aMannus replied:
It might be an idea to save these in a new array instead, just like rando already has dungeonsDone[]. The collectible flags are a vanilla thing and maintaining a rando specific list of flags might be a little difficult in comparison.

garrettjoecox replied:
I’d prefer the new array approach as well, as this approach had me worried, could you give an example on how this could be done or where it’s been done before?

aMannus replied:
As said above, take a look at how the dungeonsDone array is set/saved/used.

garrettjoecox replied:
Ah right, thanks I’ll check it out

leggettc18 replied:
If you do go this route, you'll probably need to expand the system I set up for setting pending flags on the player. Should be a simple matter of adding an additional value to the `FlagType` enum, setting that value when the player is getting a cow check, and adding a case for that value to the `Player_SetFlag` function.

garrettjoecox replied:
~Okay, so I went down this route and it seems to work temporarily; leaving room and coming back works but reloading the save it's lost, any idea if I'm missing a piece?~ nvm, fixed this

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +226,57 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+    }
+
+    if (gSaveContext.n64ddFlag && !Flags_GetCollectible(globalCtx, collectableFlag)) {
+        itemId = Randomizer_GetRandomizedItemId(GI_MILK, this->actor.id, this->actor.world.pos.x + this->actor.world.pos.z, globalCtx->sceneNum);
```
just curious, what happens if we just pass the actor params in here? are they the same for every cow in the scene?

aMannus replied:
Looking at the changes in randomizer.cpp, it does indeed look like cows have their own params. Which makes sense because the game needs to know where to place them. Would also make it not break if you ever decide to move cows again in the future.

garrettjoecox replied:
Copying from other comment: actorParams for all cows is the same, from what I can tell each cow has two actors, one is the visual cow with actorParam 0 and the second is the collision box with actorParam 1.

leggettc18 replied:
I wouldn't call this quite resolved yet. I've thought about it and x+z seems like it has pretty high potential to not be unique. As a simplified example, a cow at (10, 10) and a cow at (2, 18) will appear identical to that function. There's gotta be a better way to handle this, keep in mind that if we need to, adding properties to the cow actor is also an option.

garrettjoecox replied:
So I've still got to use the coords to differentiate two cows within the same scene, but now for the randomizer.cpp calls I'm passing in 4540 + my new cowId  as the actorParam (454 is the cow actor ID, so with 9 cows it will be 4540-4548).

Still not great, but better than passing in the X + Z as actorParam

garrettjoecox replied:
No longer calling Randomizer_GetRandomizedItemId, using RandomizerCheck now.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +226,57 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
         this->actionFunc = func_809DF730;
     } else {
-        func_8002F434(&this->actor, globalCtx, GI_MILK, 10000.0f, 100.0f);
+        func_8002F434(&this->actor, globalCtx, EnCow_GetItemFromCow(this, globalCtx, true), 10000.0f, 100.0f);
```
not sure how i feel about having the 64dd check happening instead `GetItemFromCow`, i think just having
```c
func_8002F434(&this->actor, globalCtx, gSaveContext.n64ddFlag ? EnCow_GetItemFromCow(this, globalCtx, true) : GI_MILK, 10000.0f, 100.0f);
```
might be better

i also feel like it may make sense to move `GetItemFromCow` out to randomizer.cpp (with the intention of splitting randomizer.cpp up in the future) since it should only be used when we're rando'd

briaguya0 reviewed (commented):
overall this looks like it's off to a great start!

found a couple spots that seem like they could cause issues, and a couple places that i had some questions about

aMannus commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -223,13 +285,13 @@ void func_809DF7D8(EnCow* this, GlobalContext* globalCtx) {
 
 void func_809DF870(EnCow* this, GlobalContext* globalCtx) {
     if ((Message_GetState(&globalCtx->msgCtx) == TEXT_STATE_EVENT) && Message_ShouldAdvance(globalCtx)) {
-        if (Inventory_HasEmptyBottle()) {
+        if (EnCow_GetItemFromCow(this, globalCtx, false) != ITEM_NONE) {
```
Similar to `func_8002F434()` above, I'd suggest keeping the vanilla code in and only run rando specific code when we're rando'd. Afterwards, you can rename the function to specify it's a rando function as well.

aMannus reviewed (commented):
Also added a couple of comments. Looking great though!

leggettc18 commented:
> Ice traps are broken I think, just a sound and damage animation but no ice block, no damage taken and no message box

The no message box is normal, but the no damage and no ice block is not. This may be addressed in one of the other's comments already though.

garrettjoecox committed:
Move some flags around and use global array instead of flag for shuffle cows

garrettjoecox commented:
Pushed up a few changes: 
- moved some rando checks around from feedback
- renamed a few things (`EnCow_GetItemFromCow` -> `EnCow_GetRandomizerItemFromCow`)
- moved logic for moving cows into it's own function
- created a function to identify each cow by  scene and coords `EnCow_GetCowId`
- add `cowsMilked` global array

leggettc18 commented:
>Not sure I fully understand this, might need to see a demonstration but from what I can tell it's working

So with some checks in the past we've found that if Link is in the middle of certain animations (namely pulling out or putting away his sword and/or shield, probably other items too) he will not receive the item, but the conditions for getting it all pass causing the flag to get set and the player permanently losing access to the check (unless they mess with the save editor, but even that isn't always possible). The easiest way to test this is to hold R (shield button) before the item is given, for instance during Epona's song playback (make sure you have a shield equipped for Link to actually be able to pull it out).

What the pending flag stuff in #961 actually does, is set a pending flag and flag type on the player struct which only gets set during the animation where link holds an item over his head. At the very least, this will prevent the flag from getting set before the item is actually obtained, allowing the player to try the check again. In the best case scenario, the player will get the check immediately after the blocking animation is done, if the check is given during an update function that runs every frame.

leggettc18 commented:
>add `cowsMilked` global array (this works temporarily but doesn't persist through save...)

You'll want to look at `SaveRandomizer` and `LoadRandomizerV1` in `SaveManager.cpp` for the way to fix that. You'll also need to 0 it out in the code for creating a new save in `z_sram.c` as we've done for some of the other arrays, or you'll end up with save file two having cows already milked that were milked in save file one.

garrettjoecox committed:
Fix shuffle cows bug from copy paste

garrettjoecox commented:
> You'll want to look at `SaveRandomizer` and `LoadRandomizerV1` in `SaveManager.cpp` for the way to fix that. You'll also need to 0 it out in the code for creating a new save in `z_sram.c` as we've done for some of the other arrays, or you'll end up with save file two having cows already milked that were milked in save file one.

Had it all there just a copy paste bug 👍

garrettjoecox commented:
> The easiest way to test this is to hold R (shield button) before the item is given, for instance during Epona's song playback (make sure you have a shield equipped for Link to actually be able to pull it out).

I hear the shield sound very briefly, while holding the R button after playing the song but it's interrupted by the check, is this what you mean? (notice inputs)

https://user-images.githubusercontent.com/7316699/182631017-ba4e3ed2-b227-42d2-8441-06c64d1d0e89.mov

leggettc18 commented:
I'm not able to listen to it right now but if you heard the shield sound and still got the check that means there's no issue. So you're good!

leggettc18 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3956,6 +3982,14 @@ void DrawRandoEditor(bool& open) {
+                        ImGui::Text(Settings::ShuffleCows.GetName().c_str());
+                        InsertHelpHoverText(
+                            "Enabling this will let cows give you items upon performing Epona's Song in "
+                            "front of them. There are 9 cows.");
```
Open to opinions here, but I don't think we need to specify how many cows are available, especially since there is an extra one in Master Quest which will be relevant someday.

garrettjoecox committed:
Use 454+cowId to identify cows in randomizer.cpp

garrettjoecox commented:
I've been digging but I don't think I understand enough to fix the ice trap issue, going to focus on something else for now if anyone wants to take a look, I'll come back to this in a few days probably.

briaguya0 commented:
one thing we've tended to do in soh rando is skip pretty much as much text as possible with checks, it'd be nice if instead of

play eponas -> talk to cow about the lovely song -> get item from check

we could have

play eponas -> get item from check

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,82 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
     }
 }
 
+int EnCow_GetCowId(EnCow* this, GlobalContext* globalCtx) {
```
make this return a `RadomizerCheck` instead of an int

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,82 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+
+GetItemID EnCow_GetRandomizerItemFromCow(EnCow* this, GlobalContext* globalCtx, bool setFlag) {
+    GetItemID itemId = ITEM_NONE;
+    int cowId = EnCow_GetCowId(this, globalCtx);
```
with the previous change i recommended, instead of having a `cowId` here we'd have the `RandomizerCheck` (such as `RC_KF_LINKS_HOUSE_COW`)

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,82 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+    int cowId = EnCow_GetCowId(this, globalCtx);
+
+    if (!gSaveContext.cowsMilked[cowId]) {
+        itemId = Randomizer_GetRandomizedItemId(GI_MILK, this->actor.id, 4540 + cowId, globalCtx->sceneNum);
```
then once we have the RC, instead of hacking together a value to give `Randomizer_GetRandomizedItemId`, we can just call `Randomizer_GetItemIdFromKnownCheck`

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -106,6 +109,10 @@ void EnCow_Init(Actor* thisx, GlobalContext* globalCtx) {
     EnCow* this = (EnCow*)thisx;
     s32 pad;
 
+    if (gSaveContext.n64ddFlag) {
+        EnCow_MoveCowsForRandomizer(thisx, globalCtx);
+    }
```
do we want to move the cows even when cowsanity isn't enabled?

garrettjoecox replied:
How do I go about checking this, I assume not via the `CVar_GetS32('gRandomizeShuffleCows')` since that can be changed separate from the current randomizer save's setting?

briaguya0 replied:
https://github.com/HarbourMasters/Shipwright/blob/2f5f089e7f1b88ac52d5a2c8715fbe64df07eb03/soh/soh/OTRGlobals.cpp#L1408-L1410

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -2968,10 +2968,25 @@ RandomizerCheck Randomizer::GetCheckFromActor(s16 sceneNum, s16 actorId, s16 act
                     return RC_KF_MIDOS_BOTTOM_RIGHT_CHEST;
             }
             break;
+        case 52:
```
with the other changes i suggested all of these can be removed

briaguya0 reviewed (changes requested):
after thinking about it a bit i realized all of the coords being passed in as actorParams stuff is completely unnecessary

garrettjoecox commented:
> after thinking about it a bit i realized all of the coords being passed in as actorParams stuff is completely unnecessary

I had the same thought, unfortunately I still need a 0-9 ID to store in the save, is there an easy way for the method to return a tuple? (ID, rando check ID)

briaguya0 commented:
> is there an easy way for the method to return a tuple? (ID, rando check ID)

in c++ you can use a pair, but since this is c you'd need to just manually make a struct

I'm still not fully sold on adding the cowschecked array (especially since we'd need to add that to the pendingflags logic), but once that array works with everything I'll be fine merging it in

garrettjoecox committed:
Use Randomizer_GetItemIdFromKnownCheck instead for shuffleCows

garrettjoecox committed:
Remove count of cows from tooltip

garrettjoecox committed:
Fix ice traps for cow shuffle

garrettjoecox commented:
> I'm still not fully sold on adding the cowschecked array (especially since we'd need to add that to the pendingflags logic), but once that array works with everything I'll be fine merging it in

I'm not sure I understand you here. What do I need to do with pendingflags? Everything otherwise seems to be working with the array in terms of saving/loading

garrettjoecox commented:
Alrighty ice traps are fixed, I think all/most of the feedback above has been accounted for. Ready to review again.

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -223,13 +347,21 @@ void func_809DF7D8(EnCow* this, GlobalContext* globalCtx) {
-        func_8002F434(&this->actor, globalCtx, GI_MILK, 10000.0f, 100.0f);
+        if (gSaveContext.n64ddFlag) {
+            GetItemID itemId = EnCow_GetRandomizerItemFromCow(this, globalCtx);
+            if (itemId != GI_ICE_TRAP) {
+                func_8002F434(&this->actor, globalCtx, itemId, 10000.0f, 100.0f);
+                EnCow_SetCowMilked(this, globalCtx);
+            }
```
it seems like this will make it so the cow will always ice trap the player if the randomized item is an ice trap. it seems to me we should set the cow as milked even if we get ice trapped.

garrettjoecox replied:
So this is tricky, when it's an ice trap we don't want to attempt to give it to the player yet, we want the failsafe at 216 to instead give the ice trap due to weirdness with ice traps/flags. So with that being the case, if we set the cow as milked here, when the failsafe triggers to actually set off the trap `EnCow_GetRandomizerItemFromCow` will return `ITEM_NONE`, leading to no ice trap.

Sorry it's hard to explain, even harder to understand😆

briaguya0 replied:
gotcha, so what happens if we just don't try giving anything in this function (considering the actionfunction is updated to be `func_809DF778`, why can't we just rely on that to do the item giving and ice trapping?)

garrettjoecox replied:
Great question, I think I tried this very early on and since I didn't have the ice trap issue sorted I wrote it off as not a possibility, but this works fine, updating PR

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,129 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+};
+
+struct CowInfo EnCow_GetCowInfo(EnCow* this, GlobalContext* globalCtx) {
+    s32 uniqueCoords = this->actor.world.pos.x + this->actor.world.pos.z;
```
i still don't like using x+z here, it just seems really odd. we should just check x and z separately

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,129 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+    switch (globalCtx->sceneNum) {
+        case SCENE_SOUKO: // Lon Lon Tower
+            switch (uniqueCoords) {
+                // Two cases here cause this cow is moved in randomizer
```
we shouldn't need 2 cases here because we shouldn't be calling this outside of rando

garrettjoecox replied:
For cleanliness this method is also used before the cows are moved to identify which should be moved, so I need both cases here or a separate function that would look mostly the same, opted for this.

briaguya0 replied:
but if we're not randomized we don't move cows, so why does this need to be called at all in that case?

briaguya0 replied:
oh, i think i get it, you're using this to identify the cow before moving it, and then again after moving it

it'd be nice if the comment in code made that more clear, "cause this cow is moved in randomizer" doesn't really get that point across

briaguya0 reviewed (changes requested):
definitely easier to understand now

pretty much just requesting changes because of the ice trap thing, but there are some other parts that could be cleaned up as well

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,129 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+}
+
+void EnCow_MoveCowsForRandomizer(EnCow* this, GlobalContext* globalCtx) {
+    struct CowInfo cowInfo = EnCow_GetCowInfo(this, globalCtx);
```
since there are only 2 cows being moved, i think it makes more sense to just keep the pre-moved cow identifying in here instead

garrettjoecox committed:
More changes from suggestions

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3956,6 +3959,14 @@ void DrawRandoEditor(bool& open) {
+                        // Shuffle Cows
+                        ImGui::Text(Settings::ShuffleCows.GetName().c_str());
+                        InsertHelpHoverText(
+                            "Enabling this will let cows give you items upon performing Epona's Song in "
+                            "front of them.");
```
seems this is the same text used in 3drando's description *and* in 64rando's description, but i'm still not the biggest fan of it

mostly i don't like the `Enabling this will let` part. I think i'd prefer

```
Cows give a randomized item from the pool upon performing Epona's Song in front of them.
```
or 
```
Cows give a randomized item from the pool instead of milk the first time Epona's Song is played for them.
```
or something

doesn't need to change for this PR, just felt like getting some ideas down

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,115 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+            if (this->actor.world.pos.x == -229 && this->actor.world.pos.z == 157) {
+                cowInfo.cowId = 0;
+                cowInfo.randomizerCheck = RC_LLR_TOWER_LEFT_COW;
+            } else {
```
it might be nice to do an `else if` here and check the position of the other cow too

i don't expect it to be an issue as written, but i can imagine a scenario where:
1. we update the position of the cow from line 229
2. we forget to update the values in here
3. both cows end up returning the values from the else
4. we don't notice until someone is confused when the items the cows are giving don't match their spoilerfile

briaguya0 replied:
@garrettjoecox thoughts on this one?

garrettjoecox replied:
Pushed!

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,115 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+    struct CowInfo cowInfo = EnCow_GetCowInfo(this, globalCtx);
+
+    // Only move the cow body (the tail will be moved with the body)
+    if (this->actor.params != 0) {
+        return;
+    }
```
this doesn't require `cowInfo` so we should return before getting it

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -18,6 +18,9 @@ void func_809E0070(Actor* thisx, GlobalContext* globalCtx);
 
 void func_809DF494(EnCow* this, GlobalContext* globalCtx);
 void func_809DF6BC(EnCow* this, GlobalContext* globalCtx);
+struct CowInfo EnCow_GetCowInfo(EnCow* this, GlobalContext* globalCtx);
+void EnCow_MoveCowsForRandomizer(EnCow* this, GlobalContext* globalCtx);
+GetItemID EnCow_GetRandomizerItemFromCow(EnCow* this, GlobalContext* globalCtx);
```
nitpick here, and it's probably fine either way, but since these are prefixed with `EnCow`, we probably don't need the extra cow parts in the rest of the names. I'm thinking maybe
```
EnCow_GetInfo
EnCow_MoveForRandomizer
EnCow_GetRandomizedItem
```

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -209,12 +216,115 @@ void func_809DF730(EnCow* this, GlobalContext* globalCtx) {
+    gSaveContext.cowsMilked[cowInfo.cowId] = 1;
+}
+
+GetItemID EnCow_GetRandomizerItemFromCow(EnCow* this, GlobalContext* globalCtx) {
```
not sure about the naming here

this handles the logic for
* getting the randomized item when we haven't gotten it yet
* getting milk if we have a bottle
* getting ITEM_NONE if neither of those are true

i suggested `EnCow_GetRandomizedItem` earlier, but that doesn't feel quite right. Maybe just `EnCow_GetRandomizerItem` is better but i don't really like that either.

maybe just `EnCow_GetItem` works

briaguya0 reviewed (commented):
this is looking really good! very close to ready!

left a couple comments, mostly small stuff related to naming. don't take my suggestions as prescriptions, i'm very open to any other naming ideas

garrettjoecox committed:
More suggested changes

garrettjoecox committed:
Check coords for other cows rather than falling back

briaguya0 commented:
@garrettjoecox we can probably do

> Move functions out of cow actor file?

as a part of https://github.com/HarbourMasters/Shipwright/issues/743

briaguya0 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3956,6 +3959,14 @@ void DrawRandoEditor(bool& open) {
+                        // Shuffle Cows
+                        ImGui::Text(Settings::ShuffleCows.GetName().c_str());
+                        InsertHelpHoverText(
+                            "Cows give a randomized item from the pool upon performing Epona's\n"
```
we have logic to automatically add newlines to tooltips now, so the `\n` should be removed

briaguya0 commented on soh/src/overlays/actors/ovl_En_Cow/z_en_cow.c:
```diff
@@ -18,6 +18,9 @@ void func_809E0070(Actor* thisx, GlobalContext* globalCtx);
 
 void func_809DF494(EnCow* this, GlobalContext* globalCtx);
 void func_809DF6BC(EnCow* this, GlobalContext* globalCtx);
+struct CowInfo EnCow_GetInfo(EnCow* this, GlobalContext* globalCtx);
```
this should just be `CowInfo EnCow_GetInfo` since we aren't defining the struct here, just using it

garrettjoecox replied:
Build doesn't seem to like me removing struct here

briaguya0 replied:
try moving the struct definition up? might need to use typedef

briaguya0 replied:
yeah, that was c++ brain kicking in

in c to use the struct without the keyword struct you need to typedef it

i'm fine with keeping this as is (i'd still prefer it the other way but i'm not going to let that hold back merging this)

garrettjoecox committed:
Remove newline

garrettjoecox commented:
Just a bit of feedback, I'd ease up on the reviews when it comes to naming, control flow preferences, and handling edge cases. If I didn't have to put up with reviews on a regular basis the amount of iteration required here for ~150 LoC would've totally put me off. 

It is in better shape than it would've been without, but we aren't building anything mission critical, these are cows 😄

briaguya0 commented:
> I'd ease up on the reviews when it comes to naming, control flow preferences, and handling edge cases. 

everything in rando-next will need to go through another review process before it can make it into zhora. i will need to be able to explain everything that has changed and how it is working during that review process. that PR will include a lot more than just cows, and the easier it is to tell what's happening just by reading through the code the better.

briaguya0 reviewed (approved):
i know this was a long process, thanks for sticking with it all the way through!

leggettc18 commented:
Looks like you already merged it but just wanted to chime in that everything looks OK to me here code-wise.
